crates/runtime/src/metrics/components/mod.rs

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2024-2025 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::sync::Arc;
+
+use crate::component::metrics::{MetricSpec, MetricType, MetricsProvider, ObserveMetricCallback};
+use opentelemetry::{metrics::UpDownCounter, KeyValue};
+use snafu::prelude::*;
+
+use super::{global, LazyLock, Meter};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Internal error. Please report an issue on https://github.com/spiceai/spiceai.\nMetric callback not implemented for metric {} with type {:?}", metric.name, metric.metric_type))]
+    MetricCallbackNotImplemented { metric: MetricSpec },
+
+    #[snafu(display("Internal error. Please report an issue on https://github.com/spiceai/spiceai.\nMetric {} callback has wrong type. Expected {}, got {}", metric.name, expected_type, actual_type))]
+    MetricCallbackWrongType {
+        metric: MetricSpec,
+        expected_type: &'static str,
+        actual_type: &'static str,
+    },
+}
+
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
+pub(crate) static COMPONENTS_METER: LazyLock<Meter> = LazyLock::new(|| global::meter("component"));
+
+pub(crate) static REGISTERED_COUNT: LazyLock<UpDownCounter<i64>> = LazyLock::new(|| {
+    COMPONENTS_METER
+        .i64_up_down_counter("component_metric_registered_count")
+        .with_description("Number of currently registered component metrics.")
+        .build()
+});
+
+pub(crate) fn register_component_metric(
```

Comment:
This is the logic that actually register the component metric into our broader OTel metric system. Because from OTel's perspective it looks like another metric, it automatically gets added to the `runtime.metrics` table as well.